engine/backends/proverif/proverif_backend.ml

```diff
@@ -115,35 +115,207 @@ module Print = struct
+      ( Hax_lib_protocol__crypto__AEADAlgorithm__Chacha20Poly1305,
+        fun args -> string "PLACEHOLDER_library_constructor" );
+      (* hax_lib_protocol::cal::HMACAlgorithm_HMACAlgorithm_Sha256_c *)
+      ( Hax_lib_protocol__crypto__HMACAlgorithm__Sha256,
+        fun args -> string "PLACEHOLDER_library_constructor" );
+    ]
+
+  let library_constructors_patterns :
+      (Concrete_ident_generated.name * (field_pat list -> document)) list =
+    [
+      ( Core__option__Option__Some,
+        fun args -> string "PLACEHOLDER_library_constructor" );
+      ( Core__option__Option__None,
+        fun args -> string "PLACEHOLDER_library_constructor" );
+      ( Core__ops__range__Range,
+        fun args -> string "PLACEHOLDER_library_constructor" );
+      (* hax_lib_protocol::cal::(HashAlgorithm_HashAlgorithm_Sha256_c *)
+      ( Hax_lib_protocol__crypto__HashAlgorithm__Sha256,
+        fun args -> string "PLACEHOLDER_library_constructor" );
+      (* hax_lib_protocol::cal::DHGroup_DHGroup_X25519_c *)
+      ( Hax_lib_protocol__crypto__DHGroup__X25519,
+        fun args -> string "PLACEHOLDER_library_constructor" );
+      (* hax_lib_protocol::cal::AEADAlgorithm_AEADAlgorithm_Chacha20Poly1305_c *)
+      ( Hax_lib_protocol__crypto__AEADAlgorithm__Chacha20Poly1305,
+        fun args -> string "PLACEHOLDER_library_constructor" );
+      (* hax_lib_protocol::cal::HMACAlgorithm_HMACAlgorithm_Sha256_c *)
+      ( Hax_lib_protocol__crypto__HMACAlgorithm__Sha256,
+        fun args -> string "PLACEHOLDER_library_constructor" );
+    ]
+
+  let library_types : (Concrete_ident_generated.name * document) list =
+    [
+      (* hax_lib_protocol::cal::(t_DHScalar *)
+      (Hax_lib_protocol__crypto__DHScalar, string "PLACEHOLDER_library_type");
+      (Core__option__Option, string "PLACEHOLDER_library_type");
+      (Alloc__vec__Vec, string "PLACEHOLDER_library_type");
+    ]
+
+  let assoc_known_name name (known_name, _) =
+    Global_ident.eq_name known_name name
 
   let translate_known_function fname args =
-    (List.find_exn ~f:(assoc_known_function fname) library_functions |> snd)
+    (List.find_exn ~f:(assoc_known_name fname) library_functions |> snd) args
+
+  let translate_known_constructor cname fields =
+    (List.find_exn ~f:(assoc_known_name cname) library_constructors |> snd)
+      fields
+
+  let translate_known_constructor_pat cname args =
+    (List.find_exn ~f:(assoc_known_name cname) library_constructors_patterns
+    |> snd)
       args
 
+  let translate_known_type tname =
+    List.find_exn ~f:(assoc_known_name tname) library_types |> snd
+
   let is_known_function fname =
-    List.exists ~f:(assoc_known_function fname) library_functions
+    List.exists ~f:(assoc_known_name fname) library_functions
+
+  let is_known_constructor cname =
+    List.exists ~f:(assoc_known_name cname) library_constructors
+
+  let is_known_type tname =
+    List.exists ~f:(assoc_known_name tname) library_types
```

Comment:
I know those `translate_known_*` functions are called only after calling their associated `is_known_*` functions, but I don't like to much those `find_exn`...
Maybe instead of doing `match _ with | ... where is_known_* -> code`, we should `match _ with | ... -> (match List.find ... with | Some ... -> code | _ -> None)`?

Reply:
Thanks for this suggestion, I'll try that.
Perhaps I can also refactor this more anyway, since it's all basically the same logic.

Reply:
I've simplified this a bit and removed these `find_exn`s.

Reply:
Cool, thanks!